clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java

```diff
@@ -150,10 +157,45 @@ public String toString() {
     /* Should only be used in tests. */
     public static void closeAll() {
         synchronized (LoginManager.class) {
-            for (String key : new ArrayList<>(STATIC_INSTANCES.keySet()))
+            for (LoginMetadata<String> key : new ArrayList<>(STATIC_INSTANCES.keySet()))
                 STATIC_INSTANCES.remove(key).login.close();
-            for (Password key : new ArrayList<>(DYNAMIC_INSTANCES.keySet()))
+            for (LoginMetadata<Password> key : new ArrayList<>(DYNAMIC_INSTANCES.keySet()))
                 DYNAMIC_INSTANCES.remove(key).login.close();
         }
     }
+
+    private static Class<? extends Login> loginClass(Map<String, ?> configs,
+                                                     JaasContext.Type contextType,
+                                                     String saslMechanism,
+                                                     Class<? extends Login> defaultLoginClass) {
+        String prefix  = contextType == JaasContext.Type.SERVER ? ListenerName.saslMechanismPrefix(saslMechanism) : "";
```

Comment:
From `SaslConfigs#SASL_LOGIN_CLASS_DOC`:

> For brokers, login config must be prefixed with **listener prefix** and SASL mechanism name in lower-case. For example, **listener.name.sasl_ssl**.scram-sha-256.sasl.login.class=com.example.CustomScramLogin

This code is only using the SASL mechanism name in lower-case as the prefix.  Assuming this observation is in fact a problem, I think it raises a broader issue, which is that the listener (`SASL_SSL` vs. `SASL_PLAIN`) is not known at this point in the code (or at least it isn't readily available).

Reply:
Listener prefix is removed from the config name before it gets here. This code will find `scram-sha-256.sasl.login.class` in `configs` only if it was prefixed with the listener name that this method is being invoked for. It is the same with callback handlers too.

Reply:
Ok, so no problem in the code at this point.   This behavior -- stripping the listener prefix off before passing the config map to the code when it runs for a particular listener -- probably warrants documentation, but I'm not sure where that documentation should be (or if it exists already and I just haven't seen).  Thanks for the clarification.